### ISSUE TITLE
Converting Concrete Tracer values to regular float values doesnt work

### ISSUE THREAD

**Issue by JaySandesara** [google/jax#9085]
Hello,

I am trying to convert concrete tracer values to regular float values using `astype(float)` but it still seems to return a concrete tracer value. However when I do `astype(int)` it seems to correctly convert the concrete tracer into an int. 

Here is an example:

```
import jax
from jax import numpy as jnp

def func(mu):

  print((mu.astype(float)))
  print(mu)
  print(mu.astype(int))

return mu

f = jax.grad(func)

print(f(jnp.array(5.0)))

```

This returns the following:

```
Traced<ConcreteArray(5.0)>with<JVPTrace(level=2/0)>
  with primal = Traced<ConcreteArray(5.0):JaxprTrace(level=1/0)>
       tangent = Traced<ShapedArray(float32[]):JaxprTrace(level=1/0)>
Traced<ConcreteArray(5.0)>with<JVPTrace(level=2/0)>
  with primal = Traced<ConcreteArray(5.0):JaxprTrace(level=1/0)>
       tangent = Traced<ShapedArray(float32[]):JaxprTrace(level=1/0)>
5
1.0
```

Clearly `print(mu)` and `print(mu.astype(float))` seem to return the exact same thing while `print(mu.astype(int))` returns the correct int value. Is this expected? This is a simple example but in my original function I would like to convert one of the input parameters into a regular float value.

Thanks in advance!

**Comment by jakevdp** [comment]
Thanks for the question – if anything, I think the behavior of `mu.astype(int)` returning an array rather than a tracer looks to me like a bug.

Can you say more about why you're interested in extracting values from concrete tracers? Questions like this often indicate that someone is trying to work around JAX's computational model in ways that can lead to unintended results.

**Comment by JaySandesara** [comment]
Thanks for the quick response!

So I want to use JAX to do auto-differentiation of the negative log-likelihood function which takes as input an array with the first entry of the array being the parameter `mu` and the second entry being the parameter `alpha`. Take for instance the following simple example:

```
def NLL(mu1):
    
    mu = mu1[0]                                                                                 # the first parameter mu

    alpha=mu1[1]                                                                                # the second parameter alpha

    alpha_lumi_kFact = t.obj("alpha_ATLAS_LUMI")                     # initializing a ROOT object. Essential for the NLL computation
    
    alpha_lumi_kFact.setVal(alpha)                                                 # passing alpha value to the ROOT object
    
    result = mu+alpha_lumi_kFact.getVal()                                    # simplified result for illustration purpose  

    return result
```

Without going into too much technical detail, the `setVal` method of `alpha_lumi_kFact` object in the above function takes as input only regular float values. It may not seem apparent based on the simplified example, but it is essential for my original function to set the value of `alpha_lumi_kFact` to be equal to `mu1[1]` and then do differentiation. As it is, the following gradient computation:

```
f = jax.grad(NLL)
print(f(jnp.array([1.0,0.5])))
```

gives an error:

```
---------------------------------------------------------------------------
UnfilteredStackTrace                      Traceback (most recent call last)
<ipython-input-92-7f31c8f92bd3> in <module>
      1 f = jax.grad(NLL)
----> 2 print(f(jnp.array([1.0,0.0])))

    [... skipping hidden 1 frame]

~/.local/lib/python3.6/site-packages/jax/_src/api.py in grad_f(*args, **kwargs)
    828   def grad_f(*args, **kwargs):
--> 829     _, g = value_and_grad_f(*args, **kwargs)
    830     return g

    [... skipping hidden 1 frame]

~/.local/lib/python3.6/site-packages/jax/_src/api.py in value_and_grad_f(*args, **kwargs)
    900     if not has_aux:
--> 901       ans, vjp_py = _vjp(f_partial, *dyn_args, reduce_axes=reduce_axes)
    902     else:

~/.local/lib/python3.6/site-packages/jax/_src/api.py in _vjp(fun, has_aux, reduce_axes, *primals)
   1996     out_primal, out_vjp = ad.vjp(
-> 1997         flat_fun, primals_flat, reduce_axes=reduce_axes)
   1998     out_tree = out_tree()

~/.local/lib/python3.6/site-packages/jax/interpreters/ad.py in vjp(traceable, primals, has_aux, reduce_axes)
    114   if not has_aux:
--> 115     out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
    116   else:

~/.local/lib/python3.6/site-packages/jax/interpreters/ad.py in linearize(traceable, *primals, **kwargs)
    101   jvpfun_flat, out_tree = flatten_fun(jvpfun, in_tree)
--> 102   jaxpr, out_pvals, consts = pe.trace_to_jaxpr(jvpfun_flat, in_pvals)
    103   out_primals_pvals, out_tangents_pvals = tree_unflatten(out_tree(), out_pvals)

~/.local/lib/python3.6/site-packages/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, instantiate)
    504     fun = trace_to_subjaxpr(fun, main, instantiate)
--> 505     jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
    506     assert not env

~/.local/lib/python3.6/site-packages/jax/_src/errors.py in __init__(self, tracer, context)
    148     super().__init__(
--> 149         "Abstract tracer value encountered where concrete value is expected: "
    150         f"{tracer}\n{context}{tracer._origin_msg()}\n")

UnfilteredStackTrace: AttributeError: 'str' object has no attribute '_origin_msg'

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

AttributeError                            Traceback (most recent call last)
<ipython-input-92-7f31c8f92bd3> in <module>
      1 f = jax.grad(NLL)
----> 2 print(f(jnp.array([1.0,0.0])))

~/.local/lib/python3.6/site-packages/jax/_src/errors.py in __init__(self, tracer, context)
    147   def __init__(self, tracer: "core.Tracer", context: str = ""):
    148     super().__init__(
--> 149         "Abstract tracer value encountered where concrete value is expected: "
    150         f"{tracer}\n{context}{tracer._origin_msg()}\n")
    151 

AttributeError: 'str' object has no attribute '_origin_msg'
```

The error is a bit cryptic, but I assume it is trying to say "Abstract tracer value encountered where concrete value is expected".

Note that this error goes away if I pass an explicit float value to the `alpha_lumi_kFact` object:

```
def NLL(mu1):
    
    mu = mu1[0]                                                                    # the first parameter mu

    alpha=mu1[1]                                                                  # the second parameter alpha

    alpha_lumi_kFact = t.obj("alpha_ATLAS_LUMI")
    
    alpha_lumi_kFact.setVal(0.0)                                    # setting the parameter to regular float 0.0
    
    result = mu+alpha_lumi_kFact.getVal()   

    return result
```

which then gives the correct result for the following code:

```
f = jax.grad(NLL)
print(f(jnp.array([1.0,0.0])))

[1. 1.]
```

In summary, the `setVal` method takes as input only regular float values and does not seem to work with tracer values. Any suggestions?

**Comment by jakevdp** [comment]
It sounds like you're trying to use a JAX transform on a non-JAX function: this will not work. The way JAX traces the effect of inputs on outputs for the purposes of autodiff is via these Concrete tracer values: if you extract the underlying floating point values from these tracers and compute results with them, `grad` has no way of computing the correct result.

I'd suggest either finding (or creating) a JAX implementation of the algorithm you want to use, or otherwise finding a different method of computing gradients for your code.

**Comment by JaySandesara** [comment]
Okay, this is clear to me now. Thank you for the insight!
